src/fluence/utils/utils.go

```diff
@@ -175,43 +190,43 @@ func CreateJGF(filename string, skipLabel *string) error {
 		gpuAllocatable, hasGpuAllocatable := node.Status.Allocatable["nvidia.com/gpu"]
 
 		// TODO possibly look at pod resources vs. node.Status.Allocatable
+		// Make the compute node, which is a child of the subnet
+		// The parameters here are the node name, and the parent path
+		computeNode := fluxgraph.MakeNode(node.Name, subnetNode.Metadata.Name)
 
-		workernode := fluxgraph.MakeNode(nodeIndex, false, node.Name)
-		fluxgraph.MakeEdge(subnet, workernode, jgf.ContainsRelation)
-		fluxgraph.MakeEdge(workernode, subnet, jgf.InRelation)
+		// [subnet] -> contains -> [compute node]
+		fluxgraph.MakeBidirectionalEdge(subnetNode.Id, computeNode.Id)
 
+		// Here we are adding GPU resources under nodes
 		if hasGpuAllocatable {
 			fmt.Println("GPU Resource quantity ", gpuAllocatable.Value())
 			for index := 0; index < int(gpuAllocatable.Value()); index++ {
-				gpu := fluxgraph.MakeGPU(int64(index), jgf.NvidiaGPU, 1)
-				fluxgraph.MakeEdge(workernode, gpu, jgf.ContainsRelation)
-				fluxgraph.MakeEdge(gpu, workernode, jgf.InRelation)
+
+				// The subpath (from and not including root) is the subnet -> node
+				subpath := fmt.Sprintf("%s/%s", subnetNode.Metadata.Name, computeNode.Metadata.Name)
+
+				// TODO: can this size be greater than 1?
+				gpuNode := fluxgraph.MakeGPU(jgf.NvidiaGPU, subpath, 1)
+
+				// [compute] -> contains -> [gpu]
+				fluxgraph.MakeBidirectionalEdge(computeNode.Id, gpuNode.Id)
 			}
 
 		}
 
+		// Here is where we are adding cores
 		for index := 0; index < int(availCpu); index++ {
-			core := fluxgraph.MakeCore(int64(index), jgf.CoreType)
-			fluxgraph.MakeEdge(workernode, core, jgf.ContainsRelation)
-			fluxgraph.MakeEdge(core, workernode, jgf.InRelation)
-
-			// Question from Vanessa:
-			// How can we get here and have vcores ever not equal to zero?
-			if vcores == 0 {
-				fluxgraph.MakeNFDProperties(core, int64(index), "cpu-", &node.Labels)
-			} else {
-				for virtualCore := 0; virtualCore < vcores; virtualCore++ {
-					vcore := fluxgraph.MakeVCore(core, int64(virtualCore), jgf.VirtualCoreType)
-					fluxgraph.MakeNFDProperties(vcore, int64(index), "cpu-", &node.Labels)
-				}
-			}
+			subpath := fmt.Sprintf("%s/%s", subnetNode.Metadata.Name, computeNode.Metadata.Name)
+			coreNode := fluxgraph.MakeCore(jgf.CoreType, subpath)
```

Comment:
I don't see how/whether `index` is used here. I suppose this is a larger question, but my read is that this will create cores with global core indices rather than scoped to a node. For example, if there are 2 nodes each with 10 cores in a cluster, given the code here I expect the core names in the cluster to be `core0-core19` such that `0-9` are on `node0` and `10-19` are on `node1`.

I think that's okay, but it's unconventional. Ditto for GPU and memory.

Reply:
The `index` in the loop above isn't used as an index for the graph - it's only being used for the loop. You are correct that cores (or any resource in the graph) would have a unique identifier and an indexed scoped (globally) to put them in the space of other cores. For this reason, the index for the loop here isn't what we use, but instead the resource struct that I create, associated with a resource (like core):

```go
subpath := resource.NameWithIndex()
```
That happens when we call `fluxgraph.MakeCore`

```go
// MakeCore creates a core for the graph
func (g *FluxJGF) MakeCore(name, subpath string) Node {

	// HERE sets the next index for resource "core" in the graph
	// A core is located at the subnet->node->core
	resource := g.Resources.getCounter(name, CoreType)

	// Here the full containment path will be:
	// <cluster-root>/<subnet>/<node>/<core>
	subpath = fmt.Sprintf("%s/%s", subpath, resource.NameWithIndex())
	return g.makeNewNode(resource, subpath, defaultUnit, defaultSize)
}
```

that keeps track of the core number (e.g., that might return `core0`) and can be easily called to generate the path, name, etc. I did that because I found it confusing (and redundant / error prone) to have to generate that number from some index I would pass in to a function somewhere. It also means we can have multiple loops / inputs for cores and maintain a consistent count for the resource.

Reply:
And just to clarify that the index isn't global for the graph (that is a different function call) - the getResourceCounter function [here](https://github.com/flux-framework/flux-k8s/blob/a3a4f03bbad8dd7e2acb6333e1b4fff41e91b9f1/src/fluence/jgf/types.go#L136) creates a `ResourceCounter` that derives a count that is scoped to the type (like core) and not the global index in the graph. The global count / identifier for the graph is the [ElementId](https://github.com/flux-framework/flux-k8s/blob/a3a4f03bbad8dd7e2acb6333e1b4fff41e91b9f1/src/fluence/jgf/types.go#L139).